### PR TITLE
add a `SkipNext` method on block reader

### PR DESCRIPTION
**Pull request by willscott**
alternative to #337

* [x] Confirm interface meets piece store index needs.
* [x] Tests

### CODE REVIEW

File: v2/block_reader.go [hunk 1]
```diff
@@ -122,5 +129,88 @@ func (br *BlockReader) Next() (blocks.Block, error) {
...
+// If the underlying reader used by the BlockReader is actually a ReadSeeker, this method will attempt to
+// seek over the underlying data rather than reading it into memory.
+func (br *BlockReader) SkipNext() (*BlockMetadata, error) {
+	sctSize, err := util.LdReadSize(br.r, br.opts.ZeroLengthSectionAsEOF, br.opts.MaxAllowedSectionSize)
+	if err != nil {
+		return nil, err
+	}
+
+	if sctSize == 0 {
+		_, _, err := cid.CidFromBytes([]byte{})
```
Comment (rvagg) [inline v2/block_reader.go:155]: Props for this hack to get the right error msg

File: v2/block_reader.go [hunk 2]
```diff
@@ -122,5 +129,88 @@ func (br *BlockReader) Next() (blocks.Block, error) {
...
+
+	cidSize, c, err := cid.CidFromReader(io.LimitReader(br.r, int64(sctSize)))
+	if err != nil {
+		return nil, err
+	}
+
+	blkSize := sctSize - uint64(cidSize)
+
+	if brs, ok := br.r.(io.ReadSeeker); ok {
+		if br.readerSize == -1 {
```
Comment (rvagg) [inline v2/block_reader.go:167]:
```suggestion
		// carv1 and we don't know the size, so work it out and cache it
		if br.readerSize == -1 {
```

File: v2/block_reader.go [hunk 3]
```diff
@@ -122,5 +129,88 @@ func (br *BlockReader) Next() (blocks.Block, error) {
...
+		br.offset = uint64(finalOffset)
+		return &BlockMetadata{
+			c,
+			uint64(finalOffset) - sctSize - uint64(varint.UvarintSize(sctSize)),
+			blkSize,
+		}, nil
+	}
+
+	// read to end.
+	_, err = io.CopyN(io.Discard, br.r, int64(blkSize))
```
Comment (rvagg) [inline v2/block_reader.go:201]: Any reason to not check the `written` value against expected, like we're doing for the seek version above?

**Review by rvagg — APPROVED** [review]
lgtm if it serves purpose, although I'm a little skeptical that the skipping is going to save much aside from passing slices around given the likely layering of buffered reads down to the OS and the size of our blocks being relatively small. But maybe that's worth it?

File: v2/block_reader.go (hunk 3 above)
Comment (willscott) [reply v2/block_reader.go:201]: copy-to-discard should be able to not allocate a buffer when moving the reader forward

Comment (rvagg) [reply v2/block_reader.go:202]: but it still returns a number of bytes "copied", right? so checking that the return value matches `blkSize` seems like an appropriate thing to do?

**Pull request merged**